File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maxdp=[0]*len(nums)
        mindp=[0]*len(nums)
        maxdp[0]=mindp[0]=nums[0]
        for i in range(1,len(nums)):
            t=nums[i]
            mindp[i]=min(mindp[i-1]*t,maxdp[i-1]*t,t)
            maxdp[i]=max(mindp[i-1]*t,maxdp[i-1]*t,t)
        return max(maxdp)

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums)
    {
        int[] MAX = new int[nums.length];
        int[] MIN = new int[nums.length];
        MAX[0] = nums[0];
        MIN[0] = nums[0];
        int res=nums[0];
        for(int i=1;i<nums.length;i++)
        {
            int temp=nums[i];
            MIN[i] = Math.min(Math.min(temp,temp*MIN[i-1]),temp*MAX[i-1]);
            MAX[i] = Math.max(Math.max(temp,temp*MIN[i-1]),temp*MAX[i-1]);
            res = Math.max(res,MAX[i]);
        }
        return res;
    }
}